pkg/authorization/engine/engine.go

```diff
@@ -0,0 +1,183 @@
+package engine
+
+import (
+	"math/rand"
+	"net/http"
+
+	ipapi "github.com/oauth2-proxy/oauth2-proxy/pkg/apis/ip"
+	"github.com/oauth2-proxy/oauth2-proxy/pkg/authorization"
+	"github.com/oauth2-proxy/oauth2-proxy/pkg/authorization/index"
+)
+
+// RulesEngine manages authorization Rules and which requests
+// pass or fail authorization
+type RulesEngine struct {
+	Rules   []*authorization.Rule
+	Indices []index.Index
+
+	// Whether or not to trigger indexing & rule prioritization
+	optimize bool
+
+	// Index singletons
+	pathIndex    *index.PathIndex
+	methodsIndex *index.MethodsIndex
+	ipsIndex     *index.IPsIndex
+
+	// ipParser needed for IP Rules
+	ipParser ipapi.RealClientIPParser
+}
+
+// NewRulesEngine builds a RulesEngine. It initializes the index singletons
+// as well.
+func NewRulesEngine(ipParser ipapi.RealClientIPParser) *RulesEngine {
+	return &RulesEngine{
+		Rules:   []*authorization.Rule{},
+		Indices: []index.Index{},
+
+		optimize: false,
+
+		pathIndex:    index.NewPathIndex(),
+		methodsIndex: index.NewMethodsIndex(),
+		ipsIndex:     index.NewIPsIndex(ipParser),
+
+		ipParser: ipParser,
+	}
+}
+
+// AddRule adds a authorization Rule to our RulesEngine. Rules are
+// automatically added to the appropriate Index based on their content.
+// After a threshold, optimization mode is enabled and indexing is run
+// for all requests.
+func (e *RulesEngine) AddRule(rule *authorization.Rule) {
+	e.Rules = append(e.Rules, rule)
+
+	if e.pathIndex.IndexRule(rule) {
+		e.activateIndex(e.pathIndex)
+	}
+	if e.methodsIndex.IndexRule(rule) {
+		e.activateIndex(e.methodsIndex)
+	}
+	if e.ipsIndex.IndexRule(rule) {
+		e.activateIndex(e.ipsIndex)
+	}
+
+	if len(e.Rules) > 5 {
+		e.optimize = true
+	}
+}
+
+// Allow checks if ANY Rules would allow the request
+func (e *RulesEngine) Allow(req *http.Request) bool {
+	allower := func(rule *authorization.Rule, r *http.Request) bool {
+		return rule.Allow(r, e.ipParser)
+	}
+	return e.check(req, allower)
+}
+
+// Deny checks if ANY Rules would deny the request
+func (e *RulesEngine) Deny(req *http.Request) bool {
+	denier := func(rule *authorization.Rule, r *http.Request) bool {
+		return rule.Deny(r, e.ipParser)
+	}
+	return e.check(req, denier)
+}
+
+// check compares a http.Request against our rules engine with a checker
+// function (likely a wrapper around Allow()/Deny().
+//
+// If `optimize` is false, it will perform a O(N) scan of all rules.
+// Otherwise it will use the activated Indices. The indices and rules will
+// attempt greedy reordering to place more active Rules at the front of the
+// list.
+func (e *RulesEngine) check(req *http.Request, checker func(*authorization.Rule, *http.Request) bool) bool {
+	if !e.optimize {
+		return e.checkAll(req, checker)
+	}
+
+	// Occasionally reorder the indices to get high hit rate indices first
+	if rand.Intn(100) == 1 {
+		e.prioritizeIndices()
+	}
+
+	// TODO (@NickMeves): Investigate LRU cache of Allow(req) == false results
```

Comment:
I think this will be key as far as performance to support a future vision of user potentially able to do huge RBAC management  (ie groups per path) in the proxy layer.

Likely isn't too hard to slap an LRU cache in and have the item limit be configurable by users (10000 seems like a reasonable default).

Reply:
IMHO LRU caching will only improve performance by factors, not time complexity.  Additionally it will hurt the real-time performance, when you get a cache miss you'll have particularly high latency.  I'm not even sure it's worth it.  Parallelism might be a better move honestly, we can `go ...` for each of the providers and collect the results?  It'll make the CPU spike, but should reduce the worst case performance to closer to the slowest provider configured rather than n providers * average providers.

Reply:
My bad, I should've been more clear. I was just thinking of an LRU cache for `RulesEngine.Allow(req)` results. So a `map[string]bool` like structure where I can break away from doing the RulesEngine logic for repeat HOT requests we know don't match any rules.

No actual page/content caching of any sort 😄

Reply:
Ah okay, I gotcha.  That requires assuming all rules in the `RulesEngine` are stateless though?  That might be the case now, but ?